src/Components/Server/src/Circuits/RemoteRenderer.cs

```diff
@@ -231,26 +273,37 @@ public void OnRenderCompleted(long incomingBatchId, string errorMessageOrNull)
             // synchronizes calls to hub methods. That is, it won't issue more than one call to this method from the same hub
             // at the same time on different threads.
 
-            if (!UnacknowledgedRenderBatches.TryPeek(out var nextUnacknowledgedBatch) || incomingBatchId < nextUnacknowledgedBatch.BatchId)
+            if (!_unacknowledgedRenderBatches.TryPeek(out var nextUnacknowledgedBatch) || incomingBatchId < nextUnacknowledgedBatch.BatchId)
             {
                 Log.ReceivedDuplicateBatchAck(_logger, incomingBatchId);
             }
             else
             {
                 var lastBatchId = nextUnacknowledgedBatch.BatchId;
                 // Order is important here so that we don't prematurely dequeue the last nextUnacknowledgedBatch
-                while (UnacknowledgedRenderBatches.TryPeek(out nextUnacknowledgedBatch) && nextUnacknowledgedBatch.BatchId <= incomingBatchId)
+                while (_unacknowledgedRenderBatches.TryPeek(out nextUnacknowledgedBatch) && nextUnacknowledgedBatch.BatchId <= incomingBatchId)
                 {
                     lastBatchId = nextUnacknowledgedBatch.BatchId;
-                    UnacknowledgedRenderBatches.TryDequeue(out _);
+                    // At this point the queue is no longer full, we have at least emptied one slot, so we allow a further
+                    // full queue log entry the next time it fills up.
+                    _queueIsFullNotified = false;
+                    _unacknowledgedRenderBatches.TryDequeue(out _);
                     ProcessPendingBatch(errorMessageOrNull, nextUnacknowledgedBatch);
                 }
 
                 if (lastBatchId < incomingBatchId)
                 {
                     HandleException(
                         new InvalidOperationException($"Received an acknowledgement for batch with id '{incomingBatchId}' when the last batch produced was '{lastBatchId}'."));
+                    return;
                 }
+
+                // Normally we will not have pending renders, but it might happen that we reached the limit of
+                // available buffered renders and new renders got queued.
+                // Invke ProcessBufferedRenderRequests so that we might produce any additional batch that is
+                // missing.
+                // Its also safe to use the discard as ProcessRenderQueue won't throw.
+                _ = Dispatcher.InvokeAsync(() => ProcessRenderQueue());
```

Comment:
Why do we need to dispatch here? None of the above logic is thread-safe.

Reply:
Basically you either have the sync context or you don't 👍
- If you do `InvokeAsync` should just run the code inline
- If you don't then a lot of this code seems unsafe.

Reply:
I agree no need to use `InvokeAsync`.

Also, calling `ProcessRenderQueue` does a bunch of work that is almost always unnecessary in this case. Can I suggest you put a guard clause at the top of `ProcessRenderQueue` so it just returns if `_batchBuilder.ComponentRenderQueue.Count == 0`? Then we can also remove the existing check for that condition from the bottom of `ProcessRenderQueue` itself.

Reply:
> Why do we need to dispatch here? None of the above logic is thread-safe.

Component renders must run inside the sync context, as components require a sync context when they add to the render queue, which happens as part of processing the queue. 
We also rely on the sync context to serialize render processing (meaning only 1 thread at a time operates on the queue and produces a new render). That's why its a queue and not a concurrent queue and why we don't have Interlocked.Exchange all over the place.

> * If you don't then a lot of this code seems unsafe.

Acknowledging the batch can happen outside of the sync context because of the comment mentioned above regarding signalr dispatching only 1 message a time and because the OnAfterRenderAsync calls get queued inside the sync context within ProcessRenderQueue when the UpdateDisplayAsync task doesn't complete synchronously.

> Also, calling `ProcessRenderQueue` does a bunch of work that is almost always unnecessary in this case. Can I suggest you put a guard clause at the top of `ProcessRenderQueue` so it just returns if `_batchBuilder.ComponentRenderQueue.Count == 0`? Then we can also remove the existing check for that condition from the bottom of `ProcessRenderQueue` itself.

_batchBuilder is a private member of the base class and i think it's worth exposing it. We can afford having an extra comparison on the base class, even if its not super common, I don't think it affects perf. An alternative would be to add a HasPendingRenders method/property on Renderer or a new method on Renderer that wraps the call to ProcessRenderQueue with that check.

But I don't think either of those are a better solution for simply avoiding a check.

Reply:
> Acknowledging the batch can happen outside of the sync context because of the comment mentioned above regarding signalr dispatching only 1 message a time and because the OnAfterRenderAsync calls get queued inside the sync context within ProcessRenderQueue when the UpdateDisplayAsync task doesn't complete synchronously.

Right, I see. This has confused us multiple times now, so maybe we should reconsider relying on that. If your change here means we definitely have to do an `InvokeAsync` for the `ProcessRenderQueue` call, could we instead force the entire execution of `OnRenderCompleted` to run on the sync context? That would more closely match our intuitions, and only results in the same amount of dispatching. It's not like we're really losing parallelism by doing this, since as you note, SignalR serializes the calls anyway.

> _batchBuilder is a private member of the base class and i think it's worth exposing it

Sorry if I was unclear. I meant putting the guard clause at the top of `ProcessRenderQueue` in the base class, which already has access to `_batchBuilder`.

Reply:
> Sorry if I was unclear. I meant putting the guard clause at the top of `ProcessRenderQueue` in the base class, which already has access to `_batchBuilder`.

I think I addressed this?